File: 226. Invert Binary Tree/solution.cpp
/*
4 ms
9.8 MB

TC : O(N)
SC : O(height)
*/

/**
 * @author : SahilK-027
 * @brief : Recursive solution
*/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    void solve(TreeNode*& root){
        if(root == NULL){
            return;
        }
        TreeNode* temp = root->left;
        root -> left = root->right;
        root -> right = temp;
        solve(root->right);
        solve(root->left);
    }
public:
    TreeNode* invertTree(TreeNode* root) {
        solve(root);
        return root;
    }
};

File: 226. Invert Binary Tree/solution.js
/*
75 ms
42.2 MB

TC: O(n)
SC: O(height)

@author: SahilK-027
@brief: Recursion

*/
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */
var invertTree = function(root) {
    // Base case
    if(!root){
        return null;
    }
    // Recursion will invert left subtree and right subtree
    let left = invertTree(root.left);
    let right = invertTree(root.right);
    // Our job is to invert root
    root.left = right;
    root.right = left;
    return root;
};
